src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Scopes.cs

```diff
@@ -136,9 +155,12 @@ public IPythonType GetTypeFromAnnotation(Expression expr, LookupOptions options
             }
 
             if (node.Parent != null) {
-                fromScope = gs
-                    .TraverseBreadthFirst(s => s.Children.OfType<Scope>())
-                    .FirstOrDefault(s => s.Node == node.Parent);
+                if (!_scopeLookupCache.TryGetValue(node.Parent, out fromScope)) {
+                    fromScope = gs
+                        .TraverseDepthFirst(s => s.Children.OfType<Scope>())
+                        .FirstOrDefault(s => s.Node == node.Parent);
+                    _scopeLookupCache[node.Parent] = fromScope;
```

Comment:
Can't we add the scope to the dictionary when it is created below?

Reply:
Good point